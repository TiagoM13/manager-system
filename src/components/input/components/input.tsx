import React from 'react';
import { Controller, FieldValues } from 'react-hook-form';

import { ErrorMessage } from '@/components/error-message';

import { InputProps } from './interfaces';

export const Input = <T extends FieldValues>({
  label,
  required = false,
  className = '',
  control,
  name,
  defaultValue,
  error,
  ...props
}: InputProps<T>) => {
  const classNames = `${className} w-full rounded-md border border-slate-400 py-2 px-4 outline-offset-0 outline-sky-500 text-sm text-slate-600 disabled:opacity-60`;

  return (
    <div className="w-full">
      <label htmlFor={name} className="block text-sm text-slate-600 mb-2">
        {label} {required && <span className="text-red-500">*</span>}
      </label>

      {control ? (
        <>
          <Controller
            name={name}
            control={control}
            defaultValue={defaultValue || ''}
            render={({ field }) => (
              <input {...field} {...props} className={classNames} />
            )}
          />
        </>
      ) : (
        <>
          <input {...props} className={classNames} />
        </>
      )}

      <ErrorMessage error={error?.message} />
    </div>
  );
};
