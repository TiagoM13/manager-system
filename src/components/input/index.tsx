import React from 'react';
import {
  Control,
  Controller,
  FieldError,
  FieldPath,
  FieldValues,
} from 'react-hook-form';

interface InputProps<T extends FieldValues>
  extends React.ComponentProps<'input'> {
  label?: string;
  required?: boolean;
  className?: string;
  defaultValue?: any;
  control?: Control<T>;
  name: FieldPath<T>;
  error: FieldError | undefined;
}

export const Input = <T extends FieldValues>({
  label,
  required = false,
  className = '',
  control,
  name,
  defaultValue,
  error,
  ...props
}: InputProps<T>) => {
  return (
    <div className="w-full">
      <label htmlFor={name} className="block text-sm text-slate-600 mb-2">
        {label} {required && <span className="text-red-500">*</span>}
      </label>

      <Controller
        name={name}
        control={control}
        defaultValue={defaultValue || ''}
        render={({ field }) => (
          <input
            {...field}
            {...props}
            className={`${className} w-full rounded-md border border-slate-400 py-2 px-4 outline-sky-500 text-sm text-slate-600 disabled:opacity-60`}
          />
        )}
      />

      {!!error && (
        <span className="block text-sm text-red-500 mt-1">{error.message}</span>
      )}
    </div>
  );
};
