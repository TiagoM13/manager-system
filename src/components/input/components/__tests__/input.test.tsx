import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import { Input } from '../input';

describe('<Input/>', () => {
  test('should render correctly component', () => {
    const label = 'Input Teste';
    const placeholder = 'Escreva algo';

    const { getByText, getByPlaceholderText } = render(
      <Input label={label} name="teste" placeholder={placeholder} />,
    );

    expect(getByText(label)).toBeInTheDocument();
    expect(getByPlaceholderText(placeholder)).toBeTruthy();
  });

  test('should change the input value', () => {
    const label = 'Input Teste';
    const placeholder = 'Escreva algo';

    const handleChange = jest.fn();

    const { getByPlaceholderText, asFragment } = render(
      <Input
        label={label}
        name="teste"
        placeholder={placeholder}
        onChange={handleChange}
      />,
    );

    const input = getByPlaceholderText(placeholder) as HTMLInputElement;

    fireEvent.change(input, { target: { value: 'valor' } });

    expect(input.value).toBe('valor');

    fireEvent.change(input, { target: { value: 'Novo valor' } });

    expect(input.value).toBe('Novo valor');
    expect(handleChange).toHaveBeenCalled();

    expect(asFragment()).toMatchSnapshot();
  });
});
