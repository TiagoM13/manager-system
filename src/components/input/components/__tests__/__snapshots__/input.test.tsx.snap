// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> should change the input value 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm text-slate-600 mb-2"
      for="teste"
    >
      Input Teste 
    </label>
    <input
      class=" w-full rounded-md border border-slate-400 py-2 px-4 outline-offset-0 outline-sky-500 text-sm text-slate-600 disabled:opacity-60"
      placeholder="Escreva algo"
    />
  </div>
</DocumentFragment>
`;
